design/fd-support-kcp.md

```diff
@@ -0,0 +1,111 @@
+# Failure Domain (FD) Support for Control-Plane Nodes
+
+## Summary
+
+CAPI supports failure domains (FDs) for control-plane nodes via the KCP
+controller. CAPI will look for the set of FailureDomains in the
+`ProviderCluster.Spec.FailureDomain`, if specified, the field will be copied
+to the `Cluster.Status.FailureDomains`. KCP will choose an FD from this set
+and place its value in `Machine.Spec.FailureDomain`. Currently, KCP tries
+to balance Machines evenly across all FDs.
+
+It is expected that providers will use this chosen FD (in the Machine.Spec)
+in deciding where to place the infrastructure machine. In the case of
+metal3, CAPM3 should associate the Metal3Machine with the corresponding BMH
+in the desired FD.
+
+In public clouds FailureDomains are well defined,. For example, in Azure, FDs
+map to availability zones within an azure region. For baremetal, there is no
+such mapping. However, an operator of a baremetal/metal3 cluster may still
+place their hosts across FDs in their data center or region.
+For example, two racks, connected to independent power grids, may
+constitue two FDs. Labels placed on the BMHs, by the operator, can provide a
+means to distinguish between hosts belonging to different FDs.
+
+The proposal is for CAPM3 to choose hosts based on the FD selected by KCP.
+In other words, based on the value in `Machine.Spec.FailureDomain` CAPM3 will
+choose a BMH matching the label corresponding to that FD.
+
+## User Stories
+
+As an operator who has placed their baremetal infrastructure across different
+FDs, I would like CAPM3 to associate control-plane nodes with BMHs from the
+desired FD.
+
+## Goals
+
+* Support selection of BMHs based on the desired FD for the control plane node.
+
+## Non-Goals
+
+* Failure Domains support for worker nodes (i.e. MachineDeployments).
+
+## Proposal
+
+The operator labels the BMH resource with a standard FD label. For
+example, the following standard label could be used on the BMH:
+
+```diff
+  infrastructure.cluster.x-k8s.io/failure-domain=my-fd-1
+```
+
+Today, CAPM3 chooseHost() func associates a Metal3Machine with a specific BMH
+based on labels specified in `Metal3Machine.Spec.HostSelector.MatchLabels`.
+We can expand this capability. The HostSelector field is used to narrow down
+the set of available BMHs that meet the selection criteria. When FDs are
+being utilized (i.e. `Machine.Spec.FailureDomain` != nil and Machine is for a
+control-plane node), we insert the above label into HostSelector.MatchLabels.
+This should, in effect, narrow down the set of available BMHs to only those
+in the desired FD. Additional labels in the set could further narrow down the
+set of usable BMHs.
```

Comment:
Once the standard FD label is set in BMH and desired BMH is chosen, do we remove the FD label by the controller or it is the operator's/admin's responsibility? Also, it would be very nice and easy to follow the steps if we could include example CR representations that are used in the workflow in the proposal (e.g. BMH, Machine)

Reply:
@furkatgofurov7 The "standard label" should remain on the BMH as long as the host it represents stays within the specified FD. I don't see the point it removing it. 
I'll add snippets of CRs as you asked.

Reply:
Please see "### Example Scenario" section below.